// http_otecald/index.js
var http = require("http")
var https = require("https")

var códigos_de_error = http.STATUS_CODES

var objeto_hacia_formulario = function(datos){
	var array = []
	var objeto = Object.keys(datos)
	for(var i in objeto){
		array.push([objeto[i],datos[objeto[i]]])
	}
	return array.map(function(x){return x.join("=")}).join("&")
}
var objeto_a_pedido = function(descarga){
	return [
		descarga.url
		, descarga.método
		, descarga.cabezas
		, descarga.post
		, descarga.cbs
		, descarga.poderes
	]
}
var analizar_post_datos = function(args){
	var post = (
		args.replace(
			/^([^\?]+)?([\?]+)?((&?[^=]+=[^&]+)+)/g,"$1 $2 $3 $4"
		) + " " + " "
	).split(" ")
	var post_objeto = {
		ruta: post[0]
		, datos: post[2]
	}
	return post_objeto
}
var analizar_url = function(url){
	var protocolo = "(https)?://?"
	var host = "([^/?]+)"
	var parámetros = "(.+)?"
	var regex = new RegExp("^"+protocolo+host+parámetros+"$","")
	var partes = url.replace(regex,"$1 $2 $3").split(" ")
	var post = analizar_post_datos(partes[2])
	var analizado = {
		protocolo: partes[0]
		, "sitio": partes[1]
		, "parámetros": partes[2]
		, "post_ruta": post.ruta
		, "post_datos": post.datos
	}
	return analizado
}
var descargar = function(url,método,cabezas,post,cbs,poderes){
	poderes.mostrar.log && poderes.mostrar.url && console.log(url)
	var enlace = analizar_url(url)
	var config = {
		hostname: enlace.sitio
		, port: enlace.protocolo=="https"?443:80
		, path: enlace.parámetros
		, method: método||"GET" // "GET" o "POST"
		, headers: cabezas // {"Cookie": "a=2&b=3"} cookies en forma www
	}
	if(método=="POST"){
		config.headers["Content-Type"] = "application/x-www-form-urlencoded;charset=utf-8"
		config.headers["Content-Length"] = post.length // Longitud de a=2&b=3
	}
	poderes.mostrar.log && poderes.mostrar.config && console.log(config)
	var cuerpo = ""
	var protocolo = enlace.protocolo=="https"?https:http
	var pedido = protocolo.request(config, function(respuesta) {
		if(poderes.mostrar.log){
			poderes.mostrar.pedido && console.log(pedido)
			poderes.mostrar.recurso && console.log(recurso)
			poderes.mostrar.respuesta.cabezas.$ && console.log(respuesta.headers)
			poderes.mostrar.respuesta.cabezas.post.longitud && console.log(respuesta.headers["content-length"])
			poderes.mostrar.respuesta.cabezas.post.tipo && console.log(respuesta.headers["content-type"])
			poderes.mostrar.respuesta.estado && console.log(respuesta.statusCode)
		}
		//pedido.abort()
		respuesta.on("data", function (texto) {
			cuerpo+=texto.toString()
		})
		respuesta.on("end", function(){
			poderes.mostrar.log && poderes.mostrar.cuerpo && console.log(cuerpo)
		})
	})
	if(método=="POST"){
		pedido.write(post)
	}
	pedido.on("error", function(e) {
		cbs.error(e)
		//process.exit()
		poderes.mostrar.log && poderes.mostrar.error && console.log("Error: ",e)
	})
	pedido.end()
}
var retrollamadas = function(bajar){
	var retro = {
		url: bajar.url
		, método: bajar.método || "GET"
		, cabezas: bajar.cabezas || {}
		, post: bajar.post || ""
		, cbs: bajar.cbs || {}
		, poderes: {
			mostrar: {
				log: bajar.poderes.mostrar.log || true
				, url: bajar.poderes.mostrar.url || true
				, config: bajar.poderes.mostrar.config || true
				, pedido: bajar.poderes.mostrar.pedido || true
				, recurso: bajar.poderes.mostrar.recurso || true
				, respuesta: {
					cabezas: {
						$: bajar.poderes.mostrar.respuesta.cabezas.$ || true
						, post: {
							longitud: bajar.poderes.mostrar.respuesta.cabezas.post.lognitud || false
							, tipo: bajar.poderes.mostrar.respuesta.cabezas.post.tipo || false
						}
					}
					, estado: bajar.poderes.mostrar.estado || true
				}, cuerpo: bajar.poderes.mostrar.cuerpo || false
			}
		}
	}
	return retro
}
var descargar_flavio = function(){
	console.log("Extraído de: ", "https://flaviocopes.com/node-http-post/")
	var datos = JSON.stringify({
	  todo: "Buy the milk"
	})
	console.log("Datos: ", datos, datos.length)
	var protocolo = https
	var config = {
		hostname: "flaviocopes.com",
		port: protocolo==https?443:80,
		path: "/todos",
		method: "POST",
		headers: {
			"Content-Type": "application/json"
			, "Content-Length": datos.length
		}
	}
	var pedido = protocolo.request(config,function(respuesta){
		console.log(respuesta.statusCode)

	  respuesta.on("data", function(texto){
		process.stdout.write(texto)
	  })
	})
	pedido.on("error", function(error){
	  console.error(error)
	})
	pedido.write(datos)
	pedido.end()
}
var replicar_flavio = function(){
	console.log("Extraído de: ", "https://flaviocopes.com/node-http-post/")
	var bajar = {
		url: "https://flaviocopes.com/todos"
		, método: "GET"
		, cabezas: {
			//"Cookie": "a=2&b=3"
		}, post: "todo=Buy the milk"
		, cbs: {
			error: function(){
				return true
			}
		}, poderes: {
			mostrar: {
				log: true
				, url: true
				, config: true
				, pedido: false
				, recurso: false
				, respuesta: {
					cabezas: {
						$: true
						, post: {
							longitud: true
							, tipo: true
						}
					}
					, estado: true
				}
				, cuerpo: false
			}
		}
	}
	descargar(...objeto_a_pedido(retrollamadas(bajar)))
}
module.exports = {
	códigos_de_error: códigos_de_error

	, objeto_hacia_formulario: objeto_hacia_formulario
	, objeto_a_pedido: objeto_a_pedido

	, analizar_post_datos: analizar_post_datos
	, analizar_url: analizar_url

	, descargar: descargar
	, retrollamadas: retrollamadas

	, descargar_flavio: descargar_flavio
	, replicar_flavio: replicar_flavio
}
